src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -314,17 +345,123 @@ public TypeOverloadInfo Aggregate(INamedTypeSymbol type)
+                    {
+                        correspondingBuilder = _commonItemsBuilder;
+                    }
+
+                    var item = TypeImportCompletionItem.Create(symbol, containingNamespace, _isCSharp);
+                    item.IsCached = true;
+                    correspondingBuilder.Add(new TypeImportCompletionItemInfo(item, isPublic));
+
+                    static bool IsAttributeWithAttributeSuffix(INamedTypeSymbol symbol)
+                    {
+                        // Do the symbol textual check first. Then the more expensive symbolic check.
+                        return symbol.Name.HasAttributeSuffix(isCaseSensitive: false) && symbol.IsAttribute();
+                    }
+                }
+            }
+
+            private ReferenceCacheEntry(
                 Checksum checksum,
-                ImmutableArray<TypeImportCompletionItemInfo> cachedItems)
+                bool isCSharp,
+                ImmutableArray<TypeImportCompletionItemInfo> commonItems,
+                ImmutableArray<TypeImportCompletionItemInfo> genericItems,
+                ImmutableArray<TypeImportCompletionItemInfo> attributeItems)
             {
                 Checksum = checksum;
-                CachedItems = cachedItems;
+                IsCSharp = isCSharp;
+
+                CommonItems = commonItems;
+                AttributeItems = attributeItems;
+                GenericItems = genericItems;
             }
 
+            public bool IsCSharp { get; }
+
             public Checksum Checksum { get; }
 
-            public ImmutableArray<TypeImportCompletionItemInfo> CachedItems { get; }
+            public ImmutableArray<TypeImportCompletionItemInfo> CommonItems { get; }
+
+            private ImmutableArray<TypeImportCompletionItemInfo> GenericItems { get; }
+
+            // Attribute types can't be generic
+            private ImmutableArray<TypeImportCompletionItemInfo> AttributeItems { get; }
+
+            public ImmutableArray<TypeImportCompletionItemInfo> GetGenericItems(bool isCSharp)
+            {
+                if (isCSharp == IsCSharp || GenericItems.Length == 0)
+                {
+                    return GenericItems;
+                }
+
+                // We don't want to cache this item.
+                return GenericItems.SelectAsArray(itemInfo => itemInfo.WithItem(TypeImportCompletionItem.CreateItemWithGenericDisplaySuffix(itemInfo.Item, isCSharp)));
+            }
+
+            public ImmutableArray<TypeImportCompletionItemInfo> GetAttributeItems(bool isCSharp, bool isAttributeNameContext)
+            {
+                if (!isAttributeNameContext || AttributeItems.Length == 0)
+                {
+                    return AttributeItems;
+                }
+
+                return AttributeItems.SelectAsArray(GetAppropriateAttributeItem, isCSharp);
+
+                static TypeImportCompletionItemInfo GetAppropriateAttributeItem(TypeImportCompletionItemInfo itemInfo, bool isCSharp)
```

Comment:
It located as a local function. If so, static is not required. It not, please place outside and add "private"

Reply:
Please keep 'static' when posible.  It makes it *super* easy to know that it's not capturing things.

Reply:
OK. Then, please move it out of another method :-)

---
In reply to: [294548747](https://github.com/dotnet/roslyn/pull/36395#discussion_r294548747) [](ancestors = 294548747)

Reply:
Why do you prefer private method over local function? I think a local function is more suitablein this case